### ISSUE TITLE
nscd and DynamicUser= don't mix well

### ISSUE THREAD

**Issue by arianvp** [systemd/systemd#10740]
**systemd version the issue has been seen with**
> v239

**Used distribution**
 > NixOS

**Expected behaviour you didn't see**
From the Documentation in `man systemd.exec`:

```
       User=, Group=
           (...)

           When used in conjunction with DynamicUser= the user/group name specified
           is dynamically allocated at the time the service is started, and released
           at the time the service is stopped — unless it is already allocated
           statically (see below). If DynamicUser= is not used the specified user
           and group must have been created statically in the user database no later
           than the moment the service is started, for example using the
           sysusers.d(5) facility, which is applied at boot or package install time.
```

```
$ sudo systemd-run --pty --property=DynamicUser=yes --property=User=iamatest whoami
Running as unit: run-u339.service
Press ^] three times within 1s to disconnect TTY.
iamatest
```

**Unexpected behaviour you saw**

There is no user allocated, and `whoami` fails
```
$ sudo systemd-run --pty --property=DynamicUser=yes --property=User=iamatest whoami
Running as unit: run-u339.service
Press ^] three times within 1s to disconnect TTY.
/run/current-system/sw/bin/whoami: cannot find name for user ID 63915
```

**Steps to reproduce the problem**
```
$ sudo systemd-run --pty --property=DynamicUser=yes --property=User=iamatest whoami
```

**Comment by poettering** [comment]
that suggests that your /etc/nsswitch.conf does not contain nss-systemd. Please consult the man page for details. (Quite frankly, your distro should set that up correctly for you. This is really not supposed to be user configuration, but be correct by default.)

**Comment by arianvp** [comment]
Thanks for the pointer!  Will look into that and file an issue with the Distro

**Comment by arianvp** [comment]
It does seem to be enabled. I'll investigate further.

```
[arian@t430s:~]$ cat /etc/nsswitch.conf 
passwd:    files mymachines systemd
group:     files mymachines systemd
shadow:    files

hosts:     files mymachines dns myhostname
networks:  files

ethers:    files
services:  files
protocols: files
rpc:       files
```

**Comment by poettering** [comment]
Hmm, that suggests that the nss .so for it is not installed?

**Comment by arianvp** [comment]
I straced `nscd` and it _does_ seem to load `libnss_systemd` succesfully.

```
[arian@t430s:~]$ sudo strace -E LD_LIBRARY_PATH=/nix/store/pb1vvbzzm0q9axbcbv9989mbz10f881z-systemd-239/lib  nscd -d 2>&1 | grep nss
openat(AT_FDCWD, "/etc/nsswitch.conf", O_RDONLY|O_CLOEXEC) = 4
openat(AT_FDCWD, "/nix/store/pb1vvbzzm0q9axbcbv9989mbz10f881z-systemd-239/lib/libnss_files.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/libnss_files.so.2", O_RDONLY|O_CLOEXEC) = 4
openat(AT_FDCWD, "/nix/store/pb1vvbzzm0q9axbcbv9989mbz10f881z-systemd-239/lib/libnss_mymachines.so.2", O_RDONLY|O_CLOEXEC) = 4
openat(AT_FDCWD, "/nix/store/pb1vvbzzm0q9axbcbv9989mbz10f881z-systemd-239/lib/libnss_systemd.so.2", O_RDONLY|O_CLOEXEC) = 4
openat(AT_FDCWD, "/nix/store/pb1vvbzzm0q9axbcbv9989mbz10f881z-systemd-239/lib/libnss_dns.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/libnss_dns.so.2", O_RDONLY|O_CLOEXEC) = 4
openat(AT_FDCWD, "/nix/store/pb1vvbzzm0q9axbcbv9989mbz10f881z-systemd-239/lib/libnss_myhostname.so.2", O_RDONLY|O_CLOEXEC) = 4

**Comment by arianvp** [comment]
@poettering the problem turned out to be that  `nscd` caching is the problem here.  When allocating a dynamic user,  a lookup is done in systemd,  which fails (because the user doesnt exist, and systemd is going to allocate a dynamic uid for it)  but then that answer is cached and after the dynamic user is set up, nscd will still say the user isn't created.

I'm not sure if it's systemd's task to fix this. So lets keep this closed. Just putting this out there so that if someone else faces this problem they know where to look.

**Comment by poettering** [comment]
hmm, interesting. nscd generally more a source of problems than of what it solves...

That said, for host name lookups nscd nowadays allows passing a TTL from the nss module back. I wonder if there's something like that for username lookups too. If so, we could fix that, and always declare that nscd shouldn't cache users.

**Comment by poettering** [comment]
Hmm, so I figure we could at least flush nscd's caches or so. After all nscd is still part of glibc. It's a nasty problem.

**Comment by arianvp** [comment]
A very small reproducer that is not NixOS specific:

start nscd
```
sudo systemctl start nscd
```

Create a transient unit with an open `bash` prompt:
```
sudo systemd-run --pty --property=DynamicUser=yes --property=User=arian bash
```

**in the bash prompt of the unit**:
```
$ whoami
/usr/bin/whoami:  cannot find name for user ID 63915
```

**outside the bash prompt of the unit**:
```
sudo nscd --invalidate=passwd
```
**in the bash prompt of the unit**:
```
$ whoami
iamatest
```

DBus trace of the entire interaction: https://gist.github.com/arianvp/1926442c8af8a3f001dabe3f18292a08

Especially notice that the first time `Lookup` fails:
https://gist.github.com/arianvp/1926442c8af8a3f001dabe3f18292a08#file-gistfile1-txt-L4492
https://gist.github.com/arianvp/1926442c8af8a3f001dabe3f18292a08#file-gistfile1-txt-L2496

And the second time it succeeds:
https://gist.github.com/arianvp/1926442c8af8a3f001dabe3f18292a08#file-gistfile1-txt-L10602-L10614


I agree that `nscd` causes more issues than it solves. So we're probably going to disable it in NixOS,  but `nscd` being part of `glibc` might still make this worthwhile looking into.

**Comment by poettering** [comment]
Fix waiting in #11086
